### PR TITLE
isEmpty, isNotEmpty and forEachLayer for ktx-tiled

### PR DESCRIPTION
**Pull request by Quillraven**
Some small extensions that I introduced in my own project and that might be useful for others:
- `isEmpty` and `isNotEmpty` for `MapLayers` and `MapObjects` class
- `forEachLayer` for `TiledMap` class. The purpose of this function in my case was that I needed a way to iterate over a specific type of layers. LibGDX has a `getByType` function as part of `MapLayers` class. However, this function internally always creates an array and returns it as a result or you need to pass in an array that gets filled with those layers. 
I did not like that and that is the reason for the `forEachLayer` extension. It does basically the same but without the need of an additional array object.

Let me know if this is useful or not and in case it is useful if it requires changes or not ;)

### CODE REVIEW

File: tiled/README.md [hunk 1]
```diff
@@ -215,6 +222,44 @@ map.forEachMapObject("collision") { mapObj ->
...
+
+```kotlin
+import com.badlogic.gdx.maps.MapObjects
+import com.badlogic.gdx.maps.MapLayers
+import com.badlogic.gdx.maps.tiled.TiledMap
+import ktx.tiled.*
+
+val map: TiledMap = getTiledMap()
+
+if(tiledMap.layers.isNotEmpty()) {
```
Comment (czyzby) [inline tiled/README.md:251]: Format the examples with the default KTX formatter.

File: tiled/README.md [hunk 2]
```diff
@@ -215,6 +222,44 @@ map.forEachMapObject("collision") { mapObj ->
...
+
+```kotlin
+import com.badlogic.gdx.maps.tiled.TiledMap
+import ktx.tiled.*
+
+val map: TiledMap = getTiledMap()
+
+// Iterate over all object layers and parse them.
+// Note that parseObjectLayer is only called for layers of exact type MapLayer.
+// TiledMapTileLayer which is a subclass of MapLayer will not match.
```
Comment (czyzby) [inline tiled/README.md:235]: I'd paraphrase this to make it a bit clearer:

> For example, TiledMapTileLayer - which is a subclass of MapLayer - does not have this exact type and will not match.

File: tiled/src/main/kotlin/ktx/tiled/mapLayers.kt [hunk 3]
```diff
@@ -38,3 +39,13 @@ inline fun <reified T> MapLayer.propertyOrNull(key: String): T? = properties[key
  * @return true if the property exists. Otherwise false.
  */
 fun MapLayer.containsProperty(key: String) = properties.containsKey(key)
+
+/**
+ * Returns **true** if and only if the [MapLayers] collection is empty.
+ */
+fun MapLayers.isEmpty() = this.count <= 0
```
Comment (czyzby) [inline tiled/src/main/kotlin/ktx/tiled/mapLayers.kt:46]: Is it possible for the `count` to be negative? Is that some special state that should be handled separately or just defensive programming?

File: tiled/src/main/kotlin/ktx/tiled/tiledMaps.kt [hunk 4]
```diff
@@ -153,3 +154,18 @@ inline fun TiledMap.forEachMapObject(layerName: String, action: (MapObject) -> U
...
 }
+
+/**
+ * Extension method to run a [lambda] on a specific type of [layers][MapLayer] of the [TiledMap]. The lambda
+ * takes the matching [MapLayer] as a parameter.
+ *
+ * The class matching is an exact match meaning that a given subclass like [TiledMapTileLayer] will not match
+ * for the argument [MapLayer].
+ */
+inline fun <reified T : MapLayer> TiledMap.forEachLayer(lambda: (T) -> Unit) {
```
Comment (czyzby) [inline tiled/src/main/kotlin/ktx/tiled/tiledMaps.kt:165]: Rename lambda to something along the lines of `operation` or `action`. `forEach` in stdlib uses `action`.

File: tiled/src/main/kotlin/ktx/tiled/tiledMaps.kt [hunk 5]
```diff
@@ -153,3 +154,18 @@ inline fun TiledMap.forEachMapObject(layerName: String, action: (MapObject) -> U
...
+  this.layers.forEach {
+    if (it::class == T::class) {
```
Comment (czyzby) [inline tiled/src/main/kotlin/ktx/tiled/tiledMaps.kt:167]: Is there a specific reason why you're using an exact class match? Is it like in the example with `MapLayer`, where it is used both for extension and as a concrete class for specific layers?

File: tiled/src/main/kotlin/ktx/tiled/mapObjects.kt [hunk 6]
```diff
@@ -120,3 +121,13 @@ val MapObject.shape: Shape2D
     is RectangleMapObject -> rectangle
     else -> throw MissingShapeException("MapObject of type ${this::class.java} does not have a shape.")
   }
+
+/**
+ * Returns **true** if and only if the [MapObjects] collection is empty.
+ */
+fun MapObjects.isEmpty() = this.count <= 0
```
Comment (czyzby) [inline tiled/src/main/kotlin/ktx/tiled/mapObjects.kt:128]: If it's possible for the tiled map to return a null instead of `MapObjects`, this will be a bit more useful:
```kotlin
@OptIn(ExperimentalContracts::class)
fun MapObjects?.isEmpty(): Boolean {
  contract {
    returns(false) implies (this@isEmpty != null)
  }
  return this == null || this.count <= 0
}
```
This function, along with `isNotEmpty`, would work both as a null check and a count check. If it's never null, feel free to resolve this comment. Otherwise, the same applies to `isNotEmpty` and map layers.

File: tiled/src/test/kotlin/ktx/tiled/MapLayerTest.kt [hunk 7]
```diff
@@ -44,4 +45,34 @@ class MapLayerTest {
...
+    assertTrue(actual.isEmpty())
+  }
+
+  @Test
+  fun `should return false when MapLayers is not empty`() {
+    val actual = MapLayers()
+    actual.add(MapLayer())
+
+    assertFalse(actual.isEmpty())
+  }
```
Comment (czyzby) [inline tiled/src/test/kotlin/ktx/tiled/MapLayerTest.kt:62]: This might be stretching it a bit and testing the internal implementation, but if there's a possibility of removing a layer, I'd also add a test where you add and remove the layer to see if the count is correctly updated and the extension methods work properly.

I feel like some additional tests that are borderline for LibGDX itself are a good thing, considering they don't have much of a test suite as far as I know.

**Review by czyzby — COMMENTED** [review]
Nice work overall, might require some slight adjustments.

**Comment by czyzby** [comment]
Just a heads up - I've updated the Kotlin formatter along with other dependencies, so you might want to pull the changes and reformat the code. If you're using the official formatter for the project in your IDE, you can also run `linterIdeSetup` Gradle task to update the config.

File: tiled/src/main/kotlin/ktx/tiled/mapLayers.kt (hunk 3 above)
Comment (Quillraven) [reply tiled/src/main/kotlin/ktx/tiled/mapLayers.kt:46]: It simply returns the underlying size of the array. I now just saw that there is also a `size()` function which does exactly the same and has the same comment as `getCount()`.

Shall I convert it to == 0? I always do <= 0 because it is the exact opposite of > 0. But no, the value can never be negative in my opinion.

File: tiled/src/main/kotlin/ktx/tiled/tiledMaps.kt (hunk 5 above)
Comment (Quillraven) [reply tiled/src/main/kotlin/ktx/tiled/tiledMaps.kt:167]: Tiled has four specific layer types currently:
- Tiled
- Object
- Image
- Group

For whatever reason LibGDX only implemented three of them: `MapGroupLayer`, `TiledMapImageLayer` and `TiledMapTileLayer`. And they have a generic `MapLayer` class.

The way I needed this extension was that I wanted to iterate over Object layers, get their objects and create entities out of it. Since the object layer is just a generic `MapLayer` I would get all layers again if I use an instance check.

Also, when you use `forEachLayer` you most likely want to do that for one specific layer type.

image

File: tiled/src/main/kotlin/ktx/tiled/mapObjects.kt (hunk 6 above)
Comment (Quillraven) [reply tiled/src/main/kotlin/ktx/tiled/mapObjects.kt:128]: A `Map` in LibGDX always comes with an initialized `MapLayers` and `MapProperties` object. There is no setter from what I can tell. Therefore, null should not be possible for `MapLayers` unless you hack yourself in with reflection.

The same is true for `MapLayer` and its `MapObjects` property.

Should we consider the reflection use case? I personally wouldn't do it because when you hack around like that you are on your own ;)

```Java
public class Map implements Disposable {
	private MapLayers layers = new MapLayers();
	private MapProperties properties = new MapProperties();
```

```Java
public class MapLayer {
	private String name = "";
	// ...
	private MapObjects objects = new MapObjects();
	private MapProperties properties = new MapProperties();
```

File: tiled/src/test/kotlin/ktx/tiled/MapLayerTest.kt (hunk 7 above)
Comment (Quillraven) [reply tiled/src/test/kotlin/ktx/tiled/MapLayerTest.kt:62]: I added that test case. I don't know if people ever add/remove layers manually. The way I use it is that the entire map from Tiled gets loaded with the TmxMapLoader of LibGDX. That loader prepares a TiledMap and all its content and I never manipulated the map data during runtime.

However, it is possible and maybe people out there do it, so it makes sense to have a test for that.

**Comment by Quillraven** [comment]
Pulled from the upstream and hopefully updated the formatting correctly :)

File: tiled/src/main/kotlin/ktx/tiled/mapLayers.kt (hunk 3 above)
Comment (czyzby) [reply tiled/src/main/kotlin/ktx/tiled/mapLayers.kt:46]: @Quillraven I was just curious, I wouldn't be surprised if `-1` had some special meaning.

File: tiled/src/main/kotlin/ktx/tiled/mapObjects.kt (hunk 6 above)
Comment (czyzby) [reply tiled/src/main/kotlin/ktx/tiled/mapObjects.kt:128]: > Should we consider the reflection use case? I personally wouldn't do it because when you hack around like that you are on your own ;)

No, it's OK. No need to be paranoid. :)

File: tiled/src/test/kotlin/ktx/tiled/MapLayerTest.kt (hunk 7 above)
Comment (czyzby) [reply tiled/src/test/kotlin/ktx/tiled/MapLayerTest.kt:62]: Thanks.

**Review by czyzby — CHANGES_REQUESTED** [review]
Looks good now, thanks.

**Review by czyzby — APPROVED** [review]
Misclicked the previous review. It was supposed to be an approval. :)

**Comment by Quillraven** [comment]
cool :)

**Comment by czyzby** [comment]
These additions should be available as a snapshot shortly.

@Quillraven By the way, I might be a little harsh during the reviews sometimes with the null checks and whatnot, but my goal is not to be paranoid about Java APIs. A few additional tests for edge cases are OK, but I don't want to go overboard with checks that would only make sense e.g. if reflection is involved. Some of my comments are just what-ifs to avoid shipping code with bugs under some weird conditions. I hope you're OK with that. :)

**Comment by Quillraven** [comment]
Totally fine for me and I actually think the opposite. I take it as a very valuable input and it let's me think again about some solutions.

I learned already a lot from you  :)
